File: 2023/haskell/day02b/Main.hs
module Main ( main ) where

import Data.List
import System.Environment
import System.Exit
import System.IO

type Color = String
type Amount = Int
type CubeSet = (Amount, Amount, Amount)

usage :: IO a
usage = do
  progname <- getProgName
  putStrLn $ "usage: " ++ progname ++ " <file>"
  exitWith (ExitFailure 1)

process :: String -> Int
process contents = sum powers
  where
    powers = [redNeeded * greenNeeded * blueNeeded | line <- (lines contents)
                                                   , let revealsStr = head $ drop 1 (splitOn ": " line)
                                                   , let draws = map (splitOn ", ") (splitOn "; " revealsStr)
                                                   , let (redNeeded, greenNeeded, blueNeeded) = foldl (\(r, g, b) draw ->
                                                                                                          foldl (\(r', g', b') colorAmount ->
                                                                                                                  let amountStr = head (words colorAmount)
                                                                                                                      color = head $ drop 1 (words colorAmount)
                                                                                                                      amount = read amountStr :: Int
                                                                                                                  in case color of
                                                                                                                    "red"   -> (max r' amount, g', b')
                                                                                                                    "green" -> (r', max g' amount, b')
                                                                                                                    "blue"  -> (r', g', max b' amount)
                                                                                                                    _       -> error "unknown color"
                                                                                                          ) (r, g, b) draw
                                                                                                        ) (0, 0, 0) draws
      ]

splitOn :: Eq a => [a] -> [a] -> [[a]]
splitOn _ [] = []
splitOn delimiter list =
  chunk : splitOn delimiter rest
  where
    (chunk, rest) = breakList delimiter list

breakList :: Eq a => [a] -> [a] -> ([a], [a])
breakList delimiter list =
  case stripPrefix delimiter list of
    Just suffix -> ([], suffix)
    Nothing ->
      case list of
        [] -> ([], [])
        x : xs -> let (chunk, rest) = breakList delimiter xs in (x : chunk, rest)

main :: IO ()
main = do
  args <- getArgs
  case args of
    [] -> usage
    [filename] -> do
      contents <- readFile filename
      let result = process contents
      putStrLn $ "result = " ++ show result
    _ -> usage

File: 2023/rust/day02b/src/main.rs
use std::env;
use std::fs;
use std::process;

fn usage() {
    let progname = env::args().next().unwrap();
    println!("usage: {progname} <file>");
    process::exit(1);
}

fn process(contents: &str) -> u32 {
    let mut result: u32 = 0;
    for line in contents.lines() {
        if let Some((_, reveals_str)) = line.split_once(": ") {
            let mut red_needed: u32 = 0;
            let mut green_needed: u32 = 0;
            let mut blue_needed: u32 = 0;
            for subset_str in reveals_str.split("; ") {
                for cubes_str in subset_str.split(", ") {
                    if let Some((amount_str, color)) = cubes_str.split_once(' ') {
                        let amount: u32 = amount_str.parse::<u32>().unwrap();
                        match color {
                            "red" => {
                                if amount > red_needed {
                                    red_needed = amount;
                                }
                            }
                            "green" => {
                                if amount > green_needed {
                                    green_needed = amount;
                                }
                            }
                            "blue" => {
                                if amount > blue_needed {
                                    blue_needed = amount;
                                }
                            }
                            _ => panic!("unknown color"),
                        }
                    }
                }
            }
            result += red_needed * green_needed * blue_needed;
        }
    }
    result
}

fn main() {
    if env::args().count() < 2 {
        usage();
    }
    let filename = env::args().nth(1).unwrap();
    let contents = fs::read_to_string(filename).expect("read of input file failed");
    let result = process(&contents);
    println!("result = {result}");
}
